File: GFG/POTDs/Distance of nearest cell having 1.cpp
#include <iostream>
#include <vector>
#include <queue>

class Solution {
public:
    vector<vector<int>> nearest(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        queue<pair<int, int>> q;

        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                if (grid[i][j]) {
                    q.push({i, j});
                    grid[i][j] = 0;
                } else {
                    grid[i][j] = INT_MAX;
                }
            }
        }

        vector<int> row = {-1, 1, 0, 0};
        vector<int> col = {0, 0, -1, 1};

        while (!q.empty()) {
            pair<int, int> top = q.front();
            q.pop();
            int x = top.first;
            int y = top.second;
            int time = grid[x][y];

            for (int i = 0; i < 4; ++i) {
                int newx = x + row[i];
                int newy = y + col[i];

                if (newx >= 0 && newx < n && newy >= 0 && newy < m && grid[newx][newy] == INT_MAX) {
                    grid[newx][newy] = time + 1;
                    q.push({newx, newy});
                }
            }
        }

        return grid;
}

};

File: GFG/POTDs/Distance of nearest cell having 1.py
class Solution:

    #Function to find distance of nearest 1 in the grid for each cell.
	def nearest(self, grid):
		#Code here
        # Code here
        n = len(grid)
        m = len(grid[0]) 
        q = [] 
        for i in range(n):
            for j in range(m):
                if grid[i][j]:
                    q.append([i, j])
                    grid[i][j] = 0 
                else:
                    grid[i][j] = float("inf") 
        row = [-1, 1, 0, 0]
        col = [0, 0, -1, 1] 
        while len(q): 
            top = q.pop(0)
            x = top[0]
            y = top[1]
            time = grid[x][y]
            for i in range(4):
                newx = x + row[i]
                newy = y + col[i]
                if newx >= 0 and newx < n and newy >= 0 and newy < m and grid[newx][newy] == float("inf"): 
                    grid[newx][newy] = time + 1 
                    q.append([newx, newy])
        return grid 
